app/src/main/java/io/heckel/ntfy/ui/SettingsActivity.kt

```diff
@@ -853,6 +858,91 @@ class SettingsActivity : AppCompatActivity(), PreferenceFragmentCompat.OnPrefere
+                val headersByBaseUrl = repository.getCustomHeaders()
+                    .groupBy { it.baseUrl }
+                    .map { entry ->
+                        CustomHeaderWithMetadata(entry.key, entry.value)
+                    }
+                    .sortedBy { it.baseUrl }
+
+                activity?.runOnUiThread {
+                    addCustomHeaderPreferences(headersByBaseUrl)
+                }
+            }
+        }
+
+        private fun addCustomHeaderPreferences(headersByBaseUrl: List<CustomHeaderWithMetadata>) {
+            val baseUrlsInUse = headersByBaseUrl.map { it.baseUrl }
+
+            headersByBaseUrl.forEach { serverHeaders ->
+                val baseUrl = serverHeaders.baseUrl
+                val headers = serverHeaders.headers
+
+                val preferenceCategory = PreferenceCategory(preferenceScreen.context)
+                preferenceCategory.title = shortUrl(baseUrl)
+                preferenceScreen.addPreference(preferenceCategory)
+
+                headers.forEach { header ->
+                    val preference = Preference(preferenceScreen.context)
+                    preference.title = header.name
+                    preference.summary = redactHeaderValue(header.value)
+                    preference.onPreferenceClickListener = OnPreferenceClickListener { _ ->
+                        activity?.let {
+                            CustomHeaderFragment
+                                .newInstance(header, baseUrlsInUse)
+                                .show(it.supportFragmentManager, CustomHeaderFragment.TAG)
+                        }
+                        true
+                    }
+                    preferenceCategory.addPreference(preference)
+                }
+            }
+
+            // Add header
+            val headerAddCategory = PreferenceCategory(preferenceScreen.context)
+            headerAddCategory.title = getString(R.string.settings_general_custom_headers_prefs_header_add)
+            preferenceScreen.addPreference(headerAddCategory)
+
+            val headerAddPref = Preference(preferenceScreen.context)
+            headerAddPref.title = getString(R.string.settings_general_custom_headers_prefs_header_add_title)
+            headerAddPref.summary = getString(R.string.settings_general_custom_headers_prefs_header_add_summary)
+            headerAddPref.onPreferenceClickListener = OnPreferenceClickListener { _ ->
+                activity?.let {
+                    CustomHeaderFragment
+                        .newInstance(header = null, baseUrlsInUse = baseUrlsInUse)
+                        .show(it.supportFragmentManager, CustomHeaderFragment.TAG)
+                }
+                true
+            }
+            headerAddCategory.addPreference(headerAddPref)
+        }
+
+        private fun redactHeaderValue(value: String): String {
+            return when {
+                value.isEmpty() -> "(empty)"
```

Comment:
This should never happen, right? Since you always check `headerValue.isNotEmpty()` before saving the value?